Validate submit report options

PatriciaPerozoUSDS opened the pull request:
This PR ...
This commit stops swallowing errors for invalid options and starts warning when deprecated options are submitted.
It adds unit tests for both changes and the happy path of the process function.

Test Steps:
1. Submit a report with an **invalid option** set on your dev environment. You can do so via the CLI with the following: curl -X POST -H "client:simple_report" -H "Content-Type: text/csv" --data-binary "@./examples/submission/simple_report_example.csv" "http://localhost:7071/api/reports?option=InvalidOption"
2. Submit a report with a **deprecated option** set on your dev environment. You can do so via the CLI with the following: curl -X POST -H "client:simple_report" -H "Content-Type: text/csv" --data-binary "@./examples/submission/simple_report_example.csv" "http://localhost:7071/api/reports?option=SkipInvalidItems"
3. Submit a report with a **valid option** set on your dev environment. You can do so via the CLI with the following: curl -X POST -H "client:simple_report" -H "Content-Type: text/csv" --data-binary "@./examples/submission/simple_report_example.csv" "http://localhost:7071/api/reports?option=SendImmediately"

## Changes
- Returns a 400 when invalid options are provided
- Submits report but returns a warning when a deprecated option is provided

## Checklist

### Testing
- [X] Tested locally?
- [X] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [X] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #5174 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

PatriciaPerozoUSDS commented:
I rebased this onto Seth's branch that should merge today to pre-empt the merge conflicts, thats why the base is `cle/verify-endpoint`

PatriciaPerozoUSDS commented on prime-router/src/test/kotlin/azure/ReportFunctionTests.kt:
```diff
@@ -386,4 +437,49 @@ class ReportFunctionTests {
+
+        // No report is created and an error is returned to explain the issue
+        assert(resp.status.equals(HttpStatus.BAD_REQUEST))
+        verify(exactly = 1) { actionHistory.trackLogs(any<ActionLog>()) }
```
I tried really hard to have a better validation that the error was generated here. I needed to make actionHistory a spy to mock out and prevent attempts to write to the db. Even if I didn't include line 149 `every { actionHistory.trackLogs(any<ActionLog>()) } returns Unit`, track logs was being mocked out and logs weren't being properly added even though when testing using postman I could clearly see the proper response. I'm very open to alternatives to this approach if there are ideas about a better way to test this.

PatriciaPerozoUSDS commented on prime-router/src/test/kotlin/azure/ReportFunctionTests.kt:
```diff
@@ -386,4 +437,49 @@ class ReportFunctionTests {
+
+        // Report still created but a warning is returned
+        assert(resp.status.equals(HttpStatus.CREATED))
+        verify(exactly = 1) { actionHistory.trackLogs(any<ActionLog>()) }
```
See comment on line 453

carlosfelix2 commented on prime-router/src/main/kotlin/azure/ReportFunction.kt:
```diff
@@ -140,6 +138,11 @@ class ReportFunction(
         val httpStatus: HttpStatus =
             try {
                 val options = Options.valueOfOrNone(optionsText)
+                if (Options.isDeprecated(options)) {
```
Consider changing this to `options.isDeprecated` instead of using a companion object since the operation will always be done on an instance of that enum

carlosfelix2 commented on prime-router/src/main/kotlin/azure/ReportFunction.kt:
```diff
@@ -189,6 +189,11 @@ class ReportFunction(
                     ActionLog(InvalidReportMessage(e.message ?: "Invalid request."), type = ActionLogLevel.error)
                 )
                 HttpStatus.BAD_REQUEST
+            } catch (e: Options.Companion.InvalidOptionException) {
```
I think this replaces the IllegalArgumentException and it could be removed.

PatriciaPerozoUSDS replied:
So the reason I added this is that the scope for the `IllegalArgumentException` is `report` but adding this additional catch, I can specify `parameters` as the scope of the exception.

carlosfelix2 commented on prime-router/src/main/kotlin/azure/ReportFunction.kt:
```diff
@@ -140,6 +138,11 @@ class ReportFunction(
         val httpStatus: HttpStatus =
             try {
                 val options = Options.valueOfOrNone(optionsText)
+                if (Options.isDeprecated(options)) {
```
Not related to this PR, but can there be multiple options specified???

PatriciaPerozoUSDS replied:
I don't think so which I find very confusing. It seemed out of scope to rename Options to Option for this commit but was a thought I had

carlosfelix2 commented on prime-router/src/test/kotlin/azure/ReportFunctionTests.kt:
```diff
@@ -386,4 +437,49 @@ class ReportFunctionTests {
+    @Test
+    fun `test processRequest invalid Option`() {
+        // setup steps
+        val actionHistory = spyk(ActionHistory(TaskAction.receive))
```
I think this is a good example that the function should be refactored at some point for better unit testing.  Leave as is for now.

carlosfelix2 commented on prime-router/src/main/kotlin/Report.kt:
```diff
@@ -72,12 +76,16 @@ enum class Options {
                 valueOf(input)
             } catch (ex: IllegalArgumentException) {
-                None
+                val msg = "$input is not a valid Option. Valid options: ${Options.values().joinToString()}"
```
Write a unit test to test this unit, passing valid and invalid values

carlosfelix2 commented on prime-router/src/main/kotlin/Report.kt:
```diff
@@ -72,12 +76,16 @@ enum class Options {
+                throw InvalidOptionException(msg)
             }
         }
+        fun isDeprecated(enum: Options): Boolean = enum.declaringClass.getField(enum.name)
```
Need a unit test for this unit

carlosfelix2 commented on prime-router/src/main/kotlin/SubmissionReceiver.kt:
```diff
@@ -108,6 +108,20 @@ abstract class SubmissionReceiver(
             }
         }
+
+        internal fun getSubmissionReceiver(
```
Need a unit test for this unit

PatriciaPerozoUSDS commented:
> I rebased this onto Seth's branch that should merge today to pre-empt the merge conflicts, thats why the base is `cle/verify-endpoint`

His has merged and I have removed his commits from the branch and rebased back onto master

carlosfelix2 commented on prime-router/src/main/kotlin/Report.kt:
```diff
@@ -61,9 +61,20 @@ enum class Options {
+     * Checks to see if the enum constant has an @OptionDeprecated annotation.
+     * If the annotation is present, the constant is no longer in use.
+     */
+    fun isDeprecated(): Boolean = this.declaringClass.getField(this.name)
```
One more.  To be more like Kotlin, change this to be a variable using get().  I am thinking you have not seen that yet.  The call will look like `options.isDeprecated` instead of using `()`.  Just more kotlin like.  See https://kotlinlang.org/docs/properties.html#getters-and-setters

carlosfelix2 commented on prime-router/src/main/kotlin/SubmissionReceiver.kt:
```diff
@@ -108,6 +108,27 @@ abstract class SubmissionReceiver(
         }
+
+        /**
+         * Determines what time of submission receiver to use based on [sender]
```
For kdoc, in this program the preference is to use inline arguments, so next time try to describe the arguments in the description like:
Determines what type of submission receiver to use based on the [sender] based on a given [workflowEngine] ...

carlosfelix2 replied:
Typo: change time to type

carlosfelix2 commented on prime-router/src/test/kotlin/ReportTests.kt:
```diff
@@ -745,4 +746,21 @@ class ReportTests {
+        val noneOption = Options.valueOfOrNone("None")
+        assertThat(noneOption).equals(Options.None)
+        val invalidOption = "INVALID OPTION"
+        val msg = "$invalidOption is not a valid Option. Valid options: ${Options.values().joinToString()}"
```
I will recommend against checking the text message as that can easily change and makes this test a bit brittle.  Maybe just check there is a message.

carlosfelix2 commented on prime-router/src/main/kotlin/Report.kt:
```diff
@@ -61,9 +61,20 @@ enum class Options {
+        .getAnnotation(OptionDeprecated::class.java) != null
 
     companion object {
+        class InvalidOptionException(message: String) : Exception(message)
```
Move this class out of the companion object and into the main class.

carlosfelix2 commented on prime-router/src/main/kotlin/azure/ReportFunction.kt:
```diff
@@ -139,7 +137,12 @@ class ReportFunction(
+                val option = Options.valueOfOrNone(optionsText)
+                if (option.isDeprecated()) {
+                    actionHistory.trackLogs(
+                        ActionLog(InvalidParamMessage("$optionsText is deprecated"), type = ActionLogLevel.warning)
```
Add a bit more to the text here.  When I saw the message in the API response I was wondering what kind of error was.  Maybe like "URL parameter <parameter>..."  ... Not sure about the wording, but a bit more to make it obvious what the error is.

carlosfelix2 reviewed (approved):
with some comments
